asset-registry/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::traits::EnsureOrigin;
+use frame_support::weights::constants::WEIGHT_PER_SECOND;
+use frame_support::{log, pallet_prelude::*};
+use frame_system::pallet_prelude::*;
+pub use module::*;
+use orml_traits::asset_registry::{FixedConversionRateProvider, WeightToFeeConverter};
+use orml_traits::GetByKey;
+use scale_info::TypeInfo;
+use sp_runtime::traits::AtLeast32BitUnsigned;
+use sp_runtime::traits::Bounded;
+use sp_runtime::{traits::Member, DispatchResult};
+use sp_std::prelude::*;
+use xcm::latest::prelude::*;
+use xcm_builder::TakeRevenue;
+use xcm_executor::{traits::WeightTrader, Assets};
+
+pub trait AssetProcessor<AssetId, Metadata> {
+	fn process_asset(id: Option<AssetId>, asset_metadata: &Metadata) -> Result<(AssetId, Metadata), DispatchError>;
+}
+
+#[derive(scale_info::TypeInfo, Encode, Decode, Clone, Eq, PartialEq, Debug)]
+pub struct AssetMetadata<Balance, CustomMetadata: Parameter + Member + TypeInfo> {
+	pub decimals: u32,
+	pub name: Vec<u8>,
+	pub symbol: Vec<u8>,
+	pub existential_deposit: Balance,
+	pub location: Option<MultiLocation>,
+	pub additional: CustomMetadata,
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type CustomMetadata: Parameter + Member + TypeInfo;
+
+		type AssetId: Parameter + Member + TypeInfo;
+
+		type AuthorityOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		type ProcessAsset: AssetProcessor<Self::AssetId, AssetMetadata<Self::Balance, Self::CustomMetadata>>;
+
+		/// The balance type.
+		type Balance: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Into<u128>;
+		// /// Weight information for extrinsics in this module.
+		// type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Asset was not found
+		AssetNotFound,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RegisteredAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		SetLocation {
+			asset_id: T::AssetId,
+			location: MultiLocation,
+		},
+	}
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_metadata)]
+	pub type Metadata<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<T::Balance, T::CustomMetadata>, OptionQuery>;
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_asset_id)]
+	pub type MultiLocationLookup<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
```

Comment:
We can't use xcm::latest::MultiLocation as key format here because it will change.
We can't use VersionedMultiLocation here because it will result inconsistency.
What polkadot xcm pallet doing is using VersionedMultiLocation but always force it to be the latest version and using runtime migration to keep it up to date. But that logic is super complicated.
I will say just use v2::MultiLocation for now and migrate to v3 if needed.

Reply:
Done